### PR TITLE
ENH: implement product space weighting properly

### PR DESCRIPTION
**Pull request by kohr-h**
See  919b527 for details on changes.

The major change is the factored out weighting into a new module which is re-usable by `FnBase` and `ProductSpace`. The space-specific classes simply implement `inner`, `norm` and `dist`, while much of the rest is used unchanged. Lots of saved code duplication.

~~In addition, I cleaned up some parts, mostly the import section and the doctest invocation which used to give tons of pylint errors.~~

I also moved `pspace.py` to the `space` submodule since I started to get circular deps with stuff from space (due to `FnWeightingBase` before moving that to a module on its own, but putting the latter into `set` would have been a bad fit). I think product spaces are a better fit there anyway. We could even consider moving `space.py`, but maybe that's unnecessary.

EDIT: doctest and cleanup stuff done in a separate PR #321

### CODE REVIEW

File: odl/discr/grid.py [hunk 1]
```diff
@@ -1343,5 +1341,4 @@ def uniform_sampling(begin, end, num_nodes, nodes_on_bdry=True):
 
 
 if __name__ == '__main__':
-    from doctest import testmod, NORMALIZE_WHITESPACE
-    testmod(optionflags=NORMALIZE_WHITESPACE)
+    run_doctests()
```
Comment (adler-j) [inline odl/discr/grid.py]: would really have prefered this on another branch, can we cherry pick it?

**Comment by adler-j** [comment]
This is quite hard to read due to the very extensive changes, we really should condense it to what it is (aka cherry pick 266e352).

Some design considerations: How well does these weightings generalize to non cpu spaces? Rn^2^3 type spaces?

File: odl/__init__.py [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
 from __future__ import absolute_import
 
-__version__ = '0.9b1'
+__version__ = '0.2.2'
```
Comment (adler-j) [inline odl/__init__.py]: Why is this here?

File: odl/discr/grid.py (hunk 1 above)
Comment (kohr-h) [inline odl/discr/grid.py]: Okay, okay

File: odl/__init__.py (hunk 2 above)
Comment (kohr-h) [inline odl/__init__.py]: Cherry pick.

**Comment by kohr-h** [comment]
We first merge #321, then I'll rebase, and the pull will be more readable.

**Comment by kohr-h** [comment]
> Some design considerations: How well does these weightings generalize to non cpu spaces? (Rn^2)^3 type spaces?

At least they handle the stuff we already had, including the CUDA spaces. Basically this is just a big refactor that moves all the boilerplate code out of the way and leaves the concrete weighting classes with just the job to implement the 3 essential functions. I'll double-check if there is anything in the setup that is "anti-CUDA".

**Comment by kohr-h** [comment]
Pulled out the doctest stuff.

**Comment by adler-j** [comment]
Great, will do short review, looks good overall.

File: odl/space/weighting.py [hunk 3]
```diff
@@ -0,0 +1,1047 @@
...
+                             ''.format(vector, self._vector.ndim))
+
+    @property
+    def vector(self):
+        """Weighting vector of this inner product."""
+        return self._vector
+
+    def vector_is_valid(self):
+        """Test if the vector is a valid weight, i.e. positive."""
+        return np.all(np.greater(self.vector, 0))
```
Comment (adler-j) [inline odl/space/weighting.py]: why not 

```python
np.all(self.vector > 0)
```

and where is this used externally?

File: odl/space/weighting.py [hunk 4]
```diff
@@ -0,0 +1,1047 @@
...
+        ----------
+        impl : `str`
+            Specifier for the implementation backend
+        exponent : positive `float`, optional
+            Exponent of the norm. For values other than 2.0, the inner
+            product is not defined.
+        dist_using_inner : `bool`, optional
+            Calculate `dist` using the following formula::
+
+                ||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>
```
Comment (adler-j) [inline odl/space/weighting.py]: use \`\`?

File: odl/space/weighting.py [hunk 5]
```diff
@@ -0,0 +1,1047 @@
...
+        ----------
+        x1 : `LinearSpaceVector`
+            Vector whose norm is calculated
+
+        Returns
+        -------
+        norm : `float`
+            The norm of the vector
+        """
+        return float(sqrt(self.inner(x, x).real))
```
Comment (adler-j) [inline odl/space/weighting.py]: why not simply call np.sqrt?

File: odl/space/weighting.py [hunk 6]
```diff
@@ -0,0 +1,1047 @@
...
+            dist_squared = (self.norm(x1) ** 2 + self.norm(x2) ** 2 -
+                            2 * self.inner(x1, x2).real)
+            if dist_squared < 0:  # Compensate for numerical error
+                dist_squared = 0.0
+            return float(sqrt(dist_squared))
+        else:
+            return self.norm(x1 - x2)
+
+
+class MatrixWeighting(WeightingBase):
```
Comment (adler-j) [inline odl/space/weighting.py]: It took me quite a while to figure out that this is not a implementation of `WeightingBase` but an abstract subclass, perhaps rename to `MatrixWeightingBase`?

File: odl/space/weighting.py (hunk 4 above)
Comment (kohr-h) [inline odl/space/weighting.py]: Sorry?

File: odl/space/weighting.py (hunk 6 above)
Comment (kohr-h) [inline odl/space/weighting.py]: Yeah, why not?

File: odl/space/weighting.py (hunk 5 above)
Comment (kohr-h) [inline odl/space/weighting.py]: True

File: odl/space/weighting.py (hunk 3 above)
Comment (kohr-h) [inline odl/space/weighting.py]: Not remembering your own suggestions? :-)

This version works with `CudaFnVector`.

The method is not used externally at the moment, but I see these validation functions as a debugging interface. Maybe for vectors, it's simple, but for matrices, it gets more involved.

Comment (adler-j) [inline odl/space/weighting.py]: Sure :)

Anyway, perhaps simply `is_valid` so you dont have to consider if it is a matrix or vector then?

File: odl/space/weighting.py (hunk 4 above)
Comment (adler-j) [inline odl/space/weighting.py]: Make this display as code, but perhaps the :: does that?

Comment (kohr-h) [inline odl/space/weighting.py]: That's the double colon, right. I guess it's better not to make it look like code.

File: odl/space/weighting.py (hunk 3 above)
Comment (kohr-h) [inline odl/space/weighting.py]: Sure

**Comment by adler-j** [comment]
Should be good!

**Comment by kohr-h** [comment]
Okay, there were some failing tests due to too strict check of the result (check for 7 places where computation was only in single precision). Fixed now. I'll merge after that.
